flocker/control/httpapi.py

```diff
@@ -208,6 +209,131 @@ def saved(ignored):
         saving.addCallback(saved)
         return saving
 
+    @app.route("/configuration/datasets/<dataset_id>", methods=['POST'])
+    @user_documentation(
+        """
+        Update an existing dataset.
+        """,
+        examples=[
+            u"update dataset with primary",
+        ]
+    )
+    @structured(
+        # XXX Validation of the supplied primary address is performed by input
+        # schema validation. Any further checks required?
+        inputSchema={'$ref': '/v1/endpoints.json#/definitions/datasets'},
+        outputSchema={'$ref': '/v1/endpoints.json#/definitions/datasets'},
+        schema_store=SCHEMAS
+    )
+    def update_dataset(self, dataset_id, primary=None, maximum_size=None,
+                       metadata=None):
+        """
+        Update an existing dataset in the cluster configuration.
+
+        :param unicode dataset_id: The unique identifier of the dataset.  This
+            is a string giving a UUID (per RFC 4122).
+
+        :param unicode primary: The address of the node to which the dataset
+            will be moved.
+
+        :param int maximum_size: The maximum number of bytes the dataset will
+            be capable of storing.  This may be optional or required depending
+            on the dataset backend.
+
+        :param dict metadata: A small collection of unicode key/value pairs to
+            associate with the dataset.  These items are not interpreted.  They
+            are only stored and made available for later retrieval.  Use this
+            for things like human-friendly dataset naming, ownership
+            information, etc.
+
+        :return: A ``dict`` describing the dataset which has been added to the
+            cluster configuration or giving error information if this is not
+            possible.
+        """
+        # XXX Changing volume size and metadata via the API is beyond the scope
+        # of this issue. Decide how to react.
+        # if maximum_size is None:
+        #     raise NotImplemented()
+
+        # if metadata is None:
+        #     raise NotImplemented()
+
+        # Get the current configuration.
+        deployment = self.persistence_service.get()
+
+        # Find all manifestations of requested dataset.
+        manifestations_and_nodes = list(
+            other_manifestations_from_deployment(deployment, dataset_id)
+        )
+
+        # There are no manifestations containing the requested dataset.
+        if not manifestations_and_nodes:
+            raise DATASET_NOT_FOUND
+
+        # Lookup the node that has a primary Manifestation (if any)
+        primary_manifestation, origin_node = [
+            (manifestation, node)
+            for manifestation, node in  manifestations_and_nodes
+            if manifestation.primary
+        ][0]
```

Comment:
I wrote a test to trigger an `IndexError` here, and got the following API response containing only the eliot log reference, but it's not labelled as such.

```
[FAIL]
Traceback (most recent call last):
  File "/home/richard/projects/HybridLogic/flocker/flocker/control/test/test_httpapi.py", line 486, in got_result
    result
  File "/home/richard/.virtualenvs/1342/lib/python2.7/site-packages/twisted/trial/_synctest.py", line 447, in assertEqual
    % (msg, pformat(first), pformat(second)))
twisted.trial.unittest.FailTest: not equal:
a = {u'dataset_id': u'c2cdd4d3-f757-4782-869f-cabb69b14fd4',
 u'metadata': {},
 u'primary': u'192.0.2.2'}
b = u'5afcac55-933c-4de5-9e11-1b023848b6fa,/'


flocker.control.test.test_httpapi.RealTestsUpdatePrimaryDataset.test_only_replicas

```

Perhaps I should create an issue to make this clearer?